### PR TITLE
Plfm 6508 d

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+import org.sagebionetworks.repo.model.DataType;
+import org.sagebionetworks.repo.model.EntityType;
+
+/**
+ * Basic information about an Entity combined with a user's permissions on the
+ * entity.
+ *
+ */
+public class EntityPermission {
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java:13]: Maybe UserEntityPermissions?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+public interface EntityPermissionDao {
+
+	/**
+	 * Get the entity permissions for the given user and set of entity IDs.
+	 * 
+	 * @param user
+	 * @param entityIds
+	 * @return
+	 */
+	public List<EntityPermission> getEntityPermissions(Set<Long> userGroups, List<Long> entityIds);
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java:15]: public is not needed

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package org.sagebionetworks.repo.model.dbo.entity;
+
+import java.util.List;
+import java.util.Set;
+
+public interface EntityPermissionDao {
+
+	/**
+	 * Get the entity permissions for the given user and set of entity IDs.
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java:9]: Let us fix the docs accordingly to the user -> userGroups change

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.sagebionetworks.repo.model.dbo.entity;
+
+import java.util.List;
+import java.util.Set;
+
+public interface EntityPermissionDao {
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDao.java:6]: Maybe we should stay consistent and name this UserEntityPermissionDao

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	@Autowired
+	private NamedParameterJdbcTemplate namedJdbcTemplate;
+
+	@Override
+	public List<EntityPermission> getEntityPermissions(Set<Long> userGroups, List<Long> entityIds) {
+		ValidateArgument.required(userGroups, "userGroups");
+		if (userGroups.isEmpty()) {
+			throw new IllegalArgumentException("User's groups cannot be empty");
+		}
+		ValidateArgument.required(entityIds, "entityIdss");
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImpl.java:36]: entityIds

File: lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+			(
+				SELECT N.ID, N.NODE_TYPE, N.PARENT_ID, N.ID AS BENEFACTOR_ID, A.ID AS ACL_ID, 1 AS DEPTH
+					FROM JDONODE N LEFT JOIN ACL A ON (N.ID = A.OWNER_ID AND A.OWNER_TYPE = 'ENTITY')
+                    WHERE N.ID IN (:entityIds)
+				UNION ALL
+				SELECT BEN.ENTITY_ID, BEN.ENTITY_TYPE, N.PARENT_ID, N.ID AS BENEFACTOR_ID,
+					A.ID AS ACL_ID, BEN.DEPTH + 1 AS DEPTH 
+					FROM BEN JOIN JDONODE N ON (BEN.PARENT_ID = N.ID) 
+					LEFT JOIN ACL A ON (N.ID = A.OWNER_ID AND A.OWNER_TYPE = 'ENTITY')
+					WHERE BEN.ACL_ID IS NULL AND DEPTH < 50
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql:13]: Let us use 100 as other places just for consistency

File: lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql [hunk 7]
```diff
@@ -0,0 +1,37 @@
+WITH
+	BEN AS (
+		WITH RECURSIVE BEN (ENTITY_ID, ENTITY_TYPE, PARENT_ID, BENEFACTOR_ID, ACL_ID, DEPTH) AS
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql:3]: Maybe we can add a few comments about the various blocks of SQL code

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+	Node project;
+	Node folder;
+	Node file;
+
+	Long projectId;
+	Long folderId;
+	Long fileId;
+
+	@BeforeEach
+	public void before() throws Exception {
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImplTest.java:81]: Additional test cases to include:

- acl with same id and but different object type
- two files share an acl
- add permission types that are not used

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImplTest.java [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+				BOOTSTRAP_PRINCIPAL.PUBLIC_GROUP.getPrincipalId());
+
+	}
+
+	@AfterEach
+	public void after() {
+		aclDao.truncateAll();
+		dataTypeDao.truncateAllData();
+		nodeDao.truncateAll();
+		if (userThreeId != null) {
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImplTest.java:108]: We can probably cleanup also the other principals?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+	public boolean isHasRead() {
+		return hasRead;
+	}
+
+	/**
+	 * Does the user have the read permission on this entity?
+	 * 
+	 * @param hasRead the hasRead to set
+	 */
+	public EntityPermission withtHasRead(boolean hasRead) {
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java:124]: without "t"

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+		this.dataType = dataType;
+		return this;
+	}
+
+	/**
+	 * Does the user have the read permission on this entity?
+	 * 
+	 * @return the hasRead
+	 */
+	public boolean isHasRead() {
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermission.java:115]: isHasRead -> hasRead

File: lib/models/src/main/java/org/sagebionetworks/repo/model/AuthorizationConstants.java [hunk 12]
```diff
@@ -46,6 +51,18 @@ public static boolean isBootstrapPrincipalId(long id) {
 			}
 			return false;
 		}
+		
+		/**
+		 * Get all of the bootstrap node IDs.
+		 * @return
+		 */
+		public static List<Long> getAllBootstrapIds(){
```
Comment (marcomarasca) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/AuthorizationConstants.java:59]: If we are not using it let us drop it

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.jdbc.core.RowCallbackHandler;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class EntityPermissionDaoImpl implements EntityPermissionDao {
+
+	public static final String GET_ENTITY_PERMISSION_SQL = DDLUtilsImpl
+			.loadSQLFromClasspath("sql/EntityPermissionDao/GetEntityPermissions.sql");
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/entity/EntityPermissionDaoImpl.java:25]: EntityPermissionDao might be redundant

**Review by marcomarasca — APPROVED** [review]

File: lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql (hunk 6 above)
Comment (john-hill) [reply lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql:13]: We need limit users from creating hierarchies deeper than we expected.  I feel that is a larger project that goes beyond this feature.  I created the following jira to track adding such a limit: https://sagebionetworks.jira.com/browse/PLFM-6589

Comment (marcomarasca) [reply lib/jdomodels/src/main/resources/sql/EntityPermissionDao/GetEntityPermissions.sql:13]: Agree, thanks!

**Pull request merged**